File: algorithms/[98]验证二叉搜索树/solution.py
# topics = ["树", "深度优先搜索"]

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None


from __future__ import annotations
from typing import Union


class TreeNode:
    def __init__(self, x: int):
        self.val = x
        self.left: Union[TreeNode, None] = None
        self.right: Union[TreeNode, None] = None


class Solution:
    def isValidBST(self, root: TreeNode) -> bool:
        return self.helper(root, None, None)

    def helper(self, root: Union[TreeNode, None], minVal: Union[int, None], maxVal: Union[int, None]) -> bool:
        if not root:
            return True
        if minVal is not None and root.val <= minVal:
            return False
        if maxVal is not None and root.val >= maxVal:
            return False
        return self.helper(root.left, minVal, root.val) and self.helper(root.right, root.val, maxVal)

File: algorithms/[98]验证二叉搜索树/solution.ts
// topics = ["树", "深度优先搜索", "栈"]

/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

export class TreeNode {
  val: number;
  left: TreeNode | null;
  right: TreeNode | null;
  constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
    this.val = val === undefined ? 0 : val;
    this.left = left === undefined ? null : left;
    this.right = right === undefined ? null : right;
  }
}

function isValidBST(root: TreeNode | null): boolean {
  // 二叉搜索树的中序遍历的值序列为递增序列
  // 使用栈模拟中序遍历
  const stack: (TreeNode | null)[] = [];
  let pre = -Infinity;

  while (stack.length || root !== null) {
    while (root !== null) {
      stack.push(root);
      root = root.left;
    }

    root = stack.pop()!;

    // 不满足递增，则不是二叉搜索树
    if (root.val <= pre) {
      return false;
    }

    pre = root.val;
    root = root.right;
  }

  return true;
}
